LaTeX printing: Re-establish 'plain' as default ...

mgeier opened the pull request:
... but make a few sneaky changes to make it still work.

#### References to other Issues or PRs

Unlike before, this also works when processed by LaTeX, and therefore fixes #15623.

It also closes #15329 because those changes are included.

#### Brief description of what is fixed or changed

The default `latex_mode` is now once again `'plain'`, as it was before  #15367.

For a short time the default was `'equation*'`, which isn't even plain LaTeX (it requires the `amsmath` package).
Also, `equation*` (and `equation` as well as several other environments) are typeset in the center of the page by LaTeX, which doesn't make too much sense in a REPL-like context.

To make `'plain'` great again, I changed two little things:

* The integral gets its limits below and above, with `\limits`, for all modes except `'inline'`
* When `'plain'` stuff is actually printed, it gets embedded in `$\displaystyle ...$`, which makes it left-aligned and not squished into a shallow line, see #15329, which this PR supersedes.

#### Work in Progress

I haven't changed the tests yet, I wanted to wait for some general feedback first.

I'm not sure if `use_latex='matplotlib'` works correctly. When I tried it, the result looked quite ugly, but it somehow resembled the right result. Is it supposed to look ugly?

#### Other comments

![image](https://user-images.githubusercontent.com/705404/49960358-40694300-ff10-11e8-973e-90305bd8e0bd.png)

The default should look like before, but the underlying LaTeX code is framed quite differently.

The `'equation*'` variant looks the same, but that's just because the Jupyter notebook forces it to the left. Without that, it would be center-aligned (but that's a different discussion for a later time).

The `'inline'` variant now looks actually like an inline equation (unlike before). This is of course ugly and I guess nobody ever wants to use it. But it works anyway.

It also works with `use_latex='png'`:

![image](https://user-images.githubusercontent.com/705404/49961301-a8b92400-ff12-11e8-87b7-2a3bd7d80c2f.png)

And, which is the reason I'm doing all this, it also works in `nbsphinx` using LaTeX/PDF output  (the screenshot is from the upcoming release, current `master` branch):

![image](https://user-images.githubusercontent.com/705404/49961792-ccc93500-ff13-11e8-9d12-2f12bc789bb6.png)

Note that here the `'equation*'` variant is centered, because it is actually processed by LaTeX.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* interactive
  * By default, LaTeX output in the Jupyter notebook is framed with `$\displaystyle ...$`. This prevents it from being centered in PDF output from nbconvert (see https://github.com/jupyter/notebook/issues/4060). 
<!-- END RELEASE NOTES -->

mgeier committed:
LaTeX printing: Re-establish 'plain' as default ...

... but make a few sneaky changes to make it still work.

Unlike before, this also works when processed by LaTeX,
and therefore fixes #15623.

It also closes #15329 because those changes are included.

asmeurer commented:
The matplotlib output does tend to look bad. It's mainly there for compatibility in frontends that don't support MathJax, like the qtconsole.

I don't see an issue with requiring amsmath. As far as I know, we already require it anyway. As for it being centered, that goes back to the discussion at https://github.com/jupyter/notebook/issues/4060. 

If this is merged (before the 1.4 release) we should manually update the [release notes](https://github.com/sympy/sympy/wiki/Release-Notes-for-1.4) on the wiki to remove mention of equation*, since it was changed before the last release. There is no need to mention changes in the release notes that were made and then undone before the release.

asmeurer commented:
I didn't realize the integral limits were printing differently in equation vs. plain because the Integral printer output `\limits`. I agree with the change here we should use `\limits` in plain and only not use it for inline.

asmeurer commented:
I have updated the release notes (please update the wiki if/when this is merged). I think I am +1 to these changes (after tests pass). Are there other things in the printers that make a distinction between plain and equation(*)? Are there other things that LaTeX itself prints differently there, or is it just a matter of centering and equation numbers? 

For https://github.com/jupyter/notebook/issues/4060, I think it makes sense to remove the CSS from the notebook, at least after this change is in a released version of SymPy.

mgeier commented:
> The matplotlib output does tend to look bad.

OK, then it's fine.

> It's mainly there for compatibility in frontends that don't support MathJax, like the qtconsole.

I just tried it in `qtconsole` and it seems to display the PNG version by default.
It looks nice.

> Are there other things in the printers that make a distinction between plain and equation(*)?

I grepped through the codebase and found a few cases where there are special cases for `'inline'`, those look reasonable.
But I found some strange remnants of seemingly out-of-place printing code:

https://github.com/sympy/sympy/blob/e2fbe231f48ec768bee63fb4fc54bdf9f31c3bd1/sympy/physics/vector/dyadic.py#L376-L394

https://github.com/sympy/sympy/blob/e2fbe231f48ec768bee63fb4fc54bdf9f31c3bd1/sympy/physics/vector/vector.py#L452-L470

https://github.com/sympy/sympy/blob/e2fbe231f48ec768bee63fb4fc54bdf9f31c3bd1/sympy/matrices/matrices.py#L2039-L2057

https://github.com/sympy/sympy/blob/e2fbe231f48ec768bee63fb4fc54bdf9f31c3bd1/sympy/tensor/array/ndim_array.py#L296-L314

I don't understand enough of the details of the SymPy printing system, but these seem to be mere copies of old code (including verbatim copies of a long comment), shouldn't those be superfluous?
And if they are not superfluous, they should be fixed (dollars should be removed!) and factored out into a single function, right?

I think this last one is unrelated to this issue, but I don't understand why `'inline'` is forced here:

https://github.com/sympy/sympy/blob/e2fbe231f48ec768bee63fb4fc54bdf9f31c3bd1/sympy/printing/preview.py#L184

asmeurer commented:
Matplotlib is only used if you don't have LaTeX installed. 

> but these seem to be mere copies of old code

That's actual very new code https://github.com/sympy/sympy/pull/15368. It isn't superfluous because none of those classes share a common superclass.

mgeier committed:
LaTeX printing: use $\displaystyle ...$ in a few remaining places

mgeier committed:
LaTeX preview: Use $\displaystyle ...$

mgeier committed:
LaTeX printing: update tests

mgeier commented:
> Matplotlib is only used if you don't have LaTeX installed.

OK, I see, makes sense.

> That's actual very new code #15368.

OK, "old" was probably the wrong word. I meant "wrong" in the sense that it uses double dollar signs.

> It isn't superfluous because none of those classes share a common superclass.

I see. It still seems awkward to me to have so many copies of the same code at different places. But anyway, I replaces the double dollars in all instances in de0e7f0b1d65c451ca826ad2ffff7325073682eb.

I also changed the `preview()` function from `'inline'` to using `$\displaystyle ...$` in 15259f4ebcdabcabb22f7b7292d674e7bc3f2b16.

I updated the tests (391e66e0db3bffc510c91e0f11550eebd55ace43), which now seem to pass.

Anything else?

moorepants commented:
> I see. It still seems awkward to me to have so many copies of the same code at different places.

That can be fixed if a mutable sympy abstract base class was introduced. But the ideal solution is to figure out how to make those particular classes immutable so they can then subclass from the core sympy types.

asmeurer commented:
We can put the code on a mixin object, or even just manually `Class._repr_latex_ = Basic._repr_latex_`. I didn't make a mixin class because it seemed silly to add another class to the heirarchy just for one method.

asmeurer commented:
This should have some manual testing before merging, to make sure the automatic LaTeX works and so on. Unfortunately, it is very difficult to do automatic testing for IPython because IPython keeps global state (plus their API is a bit of a pain to work with). It you want to try to add some you can, but I understand if you don't.

mgeier commented:
I did some manual tests with matrices, which worked fine, but I didn't test the vector stuff (but it should also work, since it's literally the same code change).

@asmeurer 

> it seemed silly to add another class to the heirarchy just for one method

I very much agree.

> just manually `Class._repr_latex_ = Basic._repr_latex_`

This sounds promising.
Do you want me to add this to this PR or do you want to postpone it to a later PR?

> It you want to try to add some you can

I don't know of a straightforward way to test this and TBH I don't want to get into this rabbit hole right now ...

A further idea (for this PR or for a later one) would be to create a new `mode`, e.g. something like `mode='displaystyle'`. This would avoid a bit of repetition.
What do you think about this?

asmeurer commented:
mode='displaystyle' is a good idea.

mgeier commented:
I've added a `'displaystyle'` mode in 215a179f66d289f78ec079daf3732d16986cd26d.

asmeurer commented:
It seems this breaks `init_printing(latex_mode='plain')`, because that mode doesn't return $'s.

asmeurer commented:
And something very strange happens when you do `init_printing(use_latex='png')` and then run `Add(*[Integral(1/x**i, (x, 1, 2)) for i in range(100)])`. It truncates the output to 
![image](https://user-images.githubusercontent.com/71486/50242105-931a8300-0386-11e9-9402-321d30620bcf.png). Some issue with the automatic equation wrapping?

mgeier commented:
> It seems this breaks `init_printing(latex_mode='plain')`

I would like to argue that this is actually the expected behavior.
Since `latex_mode='plain'` doesn't properly switch to LaTeX's math mode (and never did), it should not be used for `init_printing()`, because it is expected not to work.

The strange thing for me is that it *does* work for `'png'` (but only via the `'matplotlib'` fallback).

It is actually inconsistent that `'inline'` is forced for `'matplotlib'`.

If you don't agree, we should drop my last commit introducing `'displaystyle'` and keep the behavior I suggested before that (which is less consistent, but also less change to the status quo).

> And something very strange happens ...

Wow, that is indeed strange!

Interestingly, it does work fine if you use `for i in range(98)` and anything smaller than that!

Also, I would like to point out that it doesn't work too great in the current `master`, either:

![image](https://user-images.githubusercontent.com/705404/50302442-6c457500-048a-11e9-95af-17269c8863f1.png)

Technically, the PNG contains the whole expression, but it isn't quite readable, is it?

asmeurer commented:
> The strange thing for me is that it does work for 'png' (but only via the 'matplotlib' fallback).

Maybe the matplotlib code is somewhere adding dollar signs?

> Technically, the PNG contains the whole expression, but it isn't quite readable, is it?

It looks like you can double click the image and it will load it actual size and you can horizontally scroll it. Anyway, the readability of png output is on the notebook, not us. But we should definitely not produce truncated images. Actually png output is more for frontends that don't support MathJax like the qtconsole (which apparently doesn't even allow horizontal scrolling), maybe also for instances where you don't want MathJax for whatever reason.

mgeier commented:
> > The strange thing for me is that it does work for `'png'` (but only via the `'matplotlib'` fallback).
>
> Maybe the matplotlib code is somewhere adding dollar signs?

I don't think so, but before calling `_matplotlib_wrapper(s)`, the argument `mode='inline'` is forced each time. So the `mode` argument is ignored if using `'matplotlib'` including the fallback if PNG fails.
I guess this is done because `'matplotlib'` only works with `mode='inline'`, and it is more desirable to successfully display the wrong thing than to fail to display something, right?

I've investigated the "cropping mystery" a bit further. It turns out that the intermediate DVI file grows more than one page if the expression is very long. And when converting back to PNG, only the last page is used!

This is a bug in the `preview()` function. I'll try to come up with a fix for that.

This bug also appears before this PR when using `latex_mode='inline'`, you just need to use a longer expression until it flows to the second page.

mgeier commented:
Here's a fix for the aforementioned bug: #15679

asmeurer commented:
Regarding the plain thing, it is listed as one of the options for `latex_mode` in the docstring, so it should probably be supported. 

Actually now that I think about it requiring the output of latex to include the dollar signs could itself be a backwards incompatibility break, if someone is using a custom latex printer that doesn't include the dollar signs. So maybe we should keep it as it was. Was there a reason we needed to replace `'$$%s$$' % latex(...)` with just `latex(...)`?

asmeurer commented:
I guess it's because it forces two dollar signs, instead of `$\displaystyle $`. How about instead of that, we check if the string starts with a dollar sign, and if it does, leave it as is, otherwise, wrap it in double dollar signs. This makes since since in the notebook with MathJax, you don't get any math rendering at all unless you have dollar signs (or does it also support `\(`/`\[`?).

asmeurer commented:
Actually I'm confused why `equation` and `equation*` do work, because they don't include dollar signs either.

mgeier commented:
Yes, the main reason I'm doing all this is to get rid of the double dollar signs in order to move equations to the left when processed with LaTeX.

Double dollars are also discouraged for some other reason with is beyond my understanding, but many people recommend not using them in LaTeX (they seem to be an ancient TeX construct).

Many people seem to think that math *environments* also have to be enclosed by dollar signs.
But in fact, this causes an error when processed by LaTeX.
Unfortunately, MathJax is very forgiving and does indeed support that.

Whether or not something is supported in MathJax very much depends on the used configuration.

Here's the documentation:

https://docs.mathjax.org/en/latest/options/preprocessors/tex2jax.html#configure-tex2jax

By default, `processEnvironments` is `true` which means math environments are supported and there is no need for any additional markup. This is of course also `true` for the Jupyter notebook.

One thing that's not enabled by default are single dollar signs!

    inlineMath: [['\(','\)']]

But the Jupyter notebook and JupyterLab add support for single dollar signs (and the related `processEscapes` option), see https://github.com/jupyterlab/jupyterlab/blob/63090ad07a78574b1aea6ba82e1fe3ab08f9c89b/packages/mathjax2/src/index.ts#L76-L81

```js
tex2jax: {
  inlineMath: [['$', '$'], ['\\(', '\\)']],
  displayMath: [['$$', '$$'], ['\\[', '\\]']],
  processEscapes: true,
  processEnvironments: true
},
```

I've re-based this PR in order to make use of the `standalone` bugfix.

If you want to keep `'plain'` working for `init_printing()`, please just drop the last commit. I can also remove it, if you prefer.

Is there anything else I should do?

asmeurer commented:
Well MathJax is much different from an actual LaTeX environment. MathJax has to find and translate the math inside an HTML document, so it of course has to use some kind of delimiters. Jupyter actually knows that the math in question is supposed to be LaTeX, so I wonder if it could make MathJax render it even without delimiters. 

How do you feel about my suggestion to check if the output starts with `$` and I guess also `\begin` and if it doesn't, wrap it in `$$`. I'm worried not only about breaking `plain` but also any custom LaTeX printer that someone might pass in to `init_printing`.

mgeier commented:
> I wonder if it could make MathJax render it even without delimiters

You should differentiate between LaTeX "text" mode and LaTeX "math" mode.
The "delimiters" you are talking about, are to switch from "text" to "math" mode.

MathJax tries to detect those "delimiters" and then tries to display the content within them as "math".
The important thing is: even though MathJax is all about "math" mode, it still analyzes the HTML content as if it were in "text" mode, to detect any constructs that could switch to "math" mode.

Long story short, the "delimiters" are an integral part of how MathJax works, it doesn't make sense to expect MathJax to work without delimiters.

The good thing about it is that LaTeX and MathJax have a quite similar behavior, just that MathJax ignores all markup outside "math" mode, except for those constructs that switch to "math" mode (plus `\ref{}` and `\eqref{}`, and probably some other exceptions?).

The similarity gets a bit watered down because MathJax is more forgiving than LaTeX in some sense. For example, it accepts "delimiters" that switch to "math" mode, even if we are already in "math" mode. Also, it defines a few macros in "math" mode, that exist in LaTeX only in "text" mode (e.g. `\lt`, `\LaTeX`).

> How do you feel about my suggestion to check if the output starts with `$` and I guess also `\begin` and if it doesn't, wrap it in `$$`.

I think that's neither a good idea, nor necessary.

My mental model about the LaTeX output of SymPy is that it is just *LaTeX* output. It is not *LaTeX "math" mode* output!
For example, the output could be `'The result is $a^7 - \pi$.'`. This is perfectly fine LaTeX output, and only part of it is "math". IMHO, such a thing (even if it might be rare) should work for both MathJax and actual LaTeX.

Clearly, this doesn't start with a delimiter, therefore your heuristics would fail.

Also, it has the (minor) problem that the use of `$$` is discouraged.

> I'm worried not only about breaking `plain` but also any custom LaTeX printer that someone might pass in to `init_printing`.

I have no experience with custom LaTeX printers.
The docs say: "This should mimic sympy.printing.latex()".
So I would expect that they either always produce something that looks like `mode='plain'`, as this is the default of `sympy.printing.latex()`, or they should support a `mode` argument, which would also take care of things.

I don't see a problem with that, but as I said, I have no experience with those custom printers.

The only problem that I see is the name `'plain'`, because this might be confusing.
AFAIU, `'plain'` is supposed to mean "non-inline LaTeX math mode, but without the delimiters".
In this very case, we are not really dealing with "LaTeX output", but actually with "LaTeX math mode output". I think the name `'plain'` doesn't really communicate this. Sadly, I don't know a better name. Also, it's probably not a good idea to change that name now. But this could probably be clarified in the docs?

Let me re-iterate how I understand `sympy.printing.latex()`: this function is supposed to generate something that's valid in LaTeX's "text" mode, except for the case `mode='plain'`, where it is supposed to generate LaTeX code for "math" mode.

That's also the reason why `init_printing(latex_mode='plain')` doesn't strictly make a lot of sense, because it would mean the "separators" would be missing. I think disallowing `'plain'` would make more sense here.
But I could also live with the previous behavior, knowing that some "delimiters" are added implicitly (as long as the right ones are added!).

Do you agree with this view?

mgeier commented:
Regarding a custom `latex_printer` it is probably better not to add a new `mode='displaystyle'`, because using that would break existing custom LaTeX printers.

But if we drop my last commit 1cf2ca07f9f46c83f981d5b2b9f26950e20e0163, custom printers should be OK.

By default, they would be called with `mode='plain'`, they would be expected to not produce any "separators", and SymPy would add `$\displaystyle ...$` around the result.
And if any other `mode` is chosen, it should work, too, because the result is just passed along without change.

asmeurer commented:
Sorry this fell by the wayside. I'm back from break now, and I want to get this in, as it's a blocker on a 1.4 release. 

[Here](https://github.com/search?l=Jupyter+Notebook&q=init_printing+latex_printer&type=Code) are some examples of custom latex printers. The ones that use subclasses of LatexPrinter will continue to work. What would break is if someone just wrote a function that returned latex, expecting to not need to include any $$. I didn't find any examples of this on GitHub. Or if they ignore the mode argument, which would arguably be their fault because init_printing explicitly passes that argument to the custom latex function being called.

As a library function, it makes sense that `latex` should by default not include any delimiters, because the output would be include in a LaTeX document. It might even be included as part of a larger mathematical expression. 

As for the notebook, I agree that including the delimiters in the printer makes sense. My only concern here is for backwards compatibility. I don't want to break the LaTeX rendering of some notebooks. But it seems one could also make the case that any notebook that is broken isn't following the best practice of subclassing the latex printer. The docs do say that the latex_printer should "mimic sympy.latex()". 

It also breaks `latex_mode='plain'`, and isn't really clear why. 

So I think the options here are:

- Don't add mode='displaystyle'. That is, leave things in init_printing they way they are. Actually we can still add the mode to `latex` but add it manually when `latex_mode='plain'` in init_printing.

- Keep the changes as they are now and raise an exception when `latex_mode='plain'`. 

- Keep the changes as they are now but detect if the output of the latex printer doesn't have delimiters, and add them. We could also do this but raise a deprecation warning, so that it doesn't become part of the API (and we would then remove it in a few versions). See https://github.com/sympy/sympy/wiki/Deprecating-policy. 

I didn't find any explicit examples that this would break, but I do want to be careful about backwards compatibility. See https://groups.google.com/d/msg/sympy/SUX_Ugyf2ME/jn3y7759CQAJ.

asmeurer commented:
> > It also breaks latex_mode='plain', and isn't really clear why.

> I don't understand this sentence.

I mean it is confusing to users why `init_printing(latex_mode='plain')` doesn't actually render the math, when `'plain'` is a valid `mode` argument to `latex`. So we should either make it so that it works, or raise an exception on it. Note I am suggesting to raise an exception from `init_printing()`, not `latex()`.

I do like the idea of having `mode='displaystyle'` but if you think it's cleaner to remove it, that is also fine as far as I am concerned.

asmeurer commented:
By the way I had never thought of MathJax as rendering general LaTeX rather than just math mode LaTeX. I wonder if they will ever add support for non-math mode LaTeX things like tables.

mgeier commented:
OK, done.

asmeurer commented:
OK, everything looks fine to me. I did a LaTeX convert. I don't know if the Out cells are supposed to be offset like this:

<img width="752" alt="screen shot 2019-01-11 at 2 35 24 pm" src="https://user-images.githubusercontent.com/71486/51060789-24f46500-15ae-11e9-8934-8388311d71cf.png">

But now all latex is left aligned, except `equation` and `equation*` in the PDF (but they are left aligned in the notebook, not sure if that is an issue).

mgeier commented:
Thanks for merging, that's great!

Which version of `nbconvert` did you use for your example above?
I tried it with `master` and `5.4` and the "Out" prompts had the same horizontal alignment as the "In" prompts.

But the non-text output cells of `nbconvert` (when converting to LaTeX) doesn't look too great anyway, see https://github.com/jupyter/nbconvert/pull/892.
